README
#########################################################################
#   Sky Monitor for Subaru Telescope  (hskymon)
#                         Document file
#
#
#                                  ver 4.5.7    2023/08/25
#                               Copyright (C) 2003-23 Akito Tajitsu
#                                        akito.tajitsu_at_nao.ac.jp
#      Subaru Telescope, National Astronomical Observatory of Japan
#########################################################################


*** Introduction
This program (hskymon) is a GTK+2/3 based telescope monitoring 
application for SUBARU Telescope, National Astronomical Observatory
of Japan.
It can be useful for actual observation and also for obs preparation 
as a telescope simulator.
Basically, hskymon is optimized for Subaru Telescope. However,
it can be adopted to other telescopes, changing positional parameters.

For target lists to display on the program, you can input OPE
files for SOSs or CSV text lists.

The latest version (source code and Win/Mac binary packages) can be 
found  in the official page of NAOJ,
  https://www.naoj.org/Observing/tools/hskymon/
or in in GitHub,
  https://github.com/chimari/hskymon/
.
If you want to make a clone of the latest source code via GitHub, use
  git clone https://github.com/chimari/hskymon.git
in your terminal.

*** Environment
hskymon requires UNIX operating system with X window system,
Microsoft Windows (7 or later), or mac OS (Mavericks or later).

To compile this program from its source code, you need to install
Gtk+ development tools (for Windows, the maintainer is using msys2).
Either Gtk+2 (>2.24; default) or Gtk+3 (>3.22) is required.
libxml2, openssl, and json-c are also required to build the program.

In order to get telescope status, hskymon requires xmlrpc-c library.
But, Windows/Mac version does not support this function.
You need to install some libraries and developer's environment for them
to make this program.
(gtk+2-dev, xmlrpc-c3-dev, libcurl4-openssl-dev ... etc.)

If you are using an internet security software (ESET etc.), you may
have problems on HTTP/HTTPS accesses on hskymon (stacks, slowness...).
In such cases, please try to add
  C:\Program Files\Subaru Telescope\hskymon\hskymon.exe  (Win)
  /Applications/hskymon.app/Contents/MacOS/hskymon-bin   (mac)
to the exclusion list in your security software.


*** Build from the source code
  - Extract the source archive and enter to the extracted directory.

  - (only when you cloned the source tree by git)
    % touch configure configure.ac aclocal.m4 Makefile.am Makefile.in

  - % ./configure
      [option]
        --with-gtk3   : use Gtk+3 instead of Gtk+2
        --with-xmlrpc : make w/xmlrpc to get telescope status in Gen2
        --with-win32  : make for Windows (32bit/64bit)
        --with-gtkmac : make w/gtk-mac-integration for mac OS
        --with-maintain : designate deprecated functions (for maintainer)

    If you failed in configure (possiblly aclocal error),
    Please try
           % /autogen.sh
    before your configure session.

  - % make
    Then, an executable file "hskymon" will be created.

  - # make install  
    (or you can copy "hskymon" file to an appropriate directory.)


*** Command line option
     -h, --help                    : Print help message
     -i, --input [input-file]      : Set the input CSV list file
     -a, --with-allsky             : Switch on All Sky Camera
     -nt, --without-telstat        : Switch off to read Telescope Status
     -s, --server [server-address] : Override Telstat Server
     -l, --log [log-file]          : Output log file


*** XMLRPC
If you compiled hskymon with "xmlrpc" option, hskymon can get the current
telescope status from Subaru Gen2 server (effective only in the Subaru
summit network.)
In this case, you should set the address of Gen2 status server
 [1] by a commandline option "-s" (or "--server")
       % hskymon -s status.server.address.foo.bar
or
 [2] by an environment variable  "GEN2_RO_SERVER"
. The environment variable [2] yields to the command line option [1].
If either [1] or [2] is found in the start up, hskymon aborts to communicate
with the status server.


*** Input files
[1]OPE file
OPE files for SOSs can be available for objects inputs.
hskymon reads the parameter list section in OPE files likely,

   <PARAMETER_LIST>

   TGT_BD44493=OBJECT="BD+44 493" RA=022649.60 DEC=+445746.80 EQUINOX=2000.00
   TGT_HD222304=OBJECT="HD222304" RA=033908.30 DEC=+502818.20 EQUINOX=2000.00
   TGT_HD224544=OBJECT="HD224544" RA=235849.20 DEC=+322254.10 EQUINOX=2000.00
   TGT_HD225132=OBJECT="HD225132" RA=000344.40 DEC=-172009.50 EQUINOX=2000.00
           ....................................

All characters in the objectname will be capitalized.

.prm files in which standard stars are defined are also loaded
automatically.
You should write command to load *.prm files likely

*LOAD "hogehoge.prm"

in the <PARAMETER> field of OPE files.
# double quotation is required for the file name.
Current searching paths for *.prm files are
  1.  same directory with the loaded *.ope
  2.  subdirectory "COMMON/"
  3.  same level directory "../COMMON/"
  4.  "~/Procedure/"
  5.  "~/Procedure/<INST>/"
  6.  "~/Procedure/<INST>/COMMON/"
  7.  "~/Procedure/COMMON/"


As default, hskymon displays all objects defined in the loaded OPE file.
If you want to hide unused objects (in "GetObject*", "GetStandard*" or 
"SetupField*" commands in <COMMAND> field of the OPE file),
you can hide them with "Hide Objects unused in OPE file" button
(the 2nd button in ASC/Telstat menu). 

For .prm files, the objects are loaded only in the case that it is used
in commands ("GetObject*", "GetStandard*" or "SetupField*") in the OPE
file.


[2]CSV Text file
The text files written in the following CSV format are also available for
objects inputs.

	HD94028     ,105128.1 ,+201639.0 ,2000,  <comments>
	BD+36 2165  ,111248.0 ,+354343.9 ,2000,  Mv=8.4  Fe/H=-2
	HD97916     ,111554.2 ,+020512.1 ,2000,  Mv=10.2 Fe/H=-2
	BD+51 1696  ,114635.2 ,+505254.7 ,2000   
	BD-13 3442  ,114650.7 ,-140643.5 ,2000,
           ....................................


[3] Non-Sidereal Tracking File (TSC format)
Non-Sidereal tracking files for TSC can also be avairable.
Please put TSC files in an appropriate directory (maybe in the same 
directory where your OPE files locate) and load them from the menu.
Files must follow the format described in
  https://www.naoj.org/Observing/Telescope/Tracking/NonSidereal/
. You should use "Geocentric" coordinate in TSC files.
They would be automatically converted into topocentric in the program.
When hskymon's time and date are out of the range of the tracking 
calculation, the object will be plotted with an "x" mark in the monitor.
You can check the orbit of the non-sidereal target in the Finsing Chart.


[4] Non-Sidereal Tracking File (JPL HORIZONS format)
hskymon can also load NASA JPL HIRIZONS output files.
  https://ssd.jpl.nasa.gov/horizons.cgi
Please note that you should set 
   "Epehmeris Type" = OBSERVER
   "Observer Location" = Geocentric
   "Table Settings" = 1. Astrometric RA & Dec, 20. Observer range
for your input.
Either calendar date or JD is acceptable for date format.
From the main menu, HORIZONS files can be converted into the TSC format.


*** Copyright
   The code in this distribution is Copyright 2003-2018 by Akito Tajitsu.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

   In addition, as a special exception, A.Tajitsu gives permission to link 
   this code of this program with the OpenSSL library (or with modified
   versions of OpenSSL that use the same license as OpenSSL), and distribute
   linked combinations including the two. You must obey the GNU General 
   Public License in all respects for all of the code used other than OpenSSL.
   If you modify this file, you may extend this exception to your version of 
   the file, but you are not obligated to do so. If you do not wish to do so,
   delete this exception statement from your version.


ChangeLog
2023.08.25 ver4.5.7 - Bug fix for SDSS FCDB.

2023.08.25 ver4.5.6 - NonBlocking recv and SSL_read for all HTTP/HTTPS accesses.
                    - Catalog updates for FC (SDSS DR18) and FCDB (SDSS DR16,
	              GSC, 2MASS etc.)

2023.08.24 ver4.5.5 - NonBlocking recv (only for HTTP access).

2023.08.23 ver4.5.4 - Test for macOS recv() packet size #2.

2023.08.23 ver4.5.3 - Test for macOS recv() packet size.

2023.08.22 ver4.5.2 - Bug fix on parameter initializtion for HTTP accesses.

2023.03.01 ver4.5.1 - Added a new function to access to Seimei Telescope
	              Status Server (only w/o xmlrpc).
	            - Added HTTP / HTTPS connection timeout.

2023.02.22 ver4.5.0 - Proxy access for all HTTP/HTTPS connections.

2023.02.14 ver4.4.1 - Bug fix for compling w/Ubuntu 22.04LTS (GCC10).
                    - Updated accessing to databases
	                 HST archive
	                 Kepler archive
	                 DSS via STSci
	                 DSS via ESO arhive
	                 LAMOST DR7 & DR8
	            - Updated accessing to FC image servers
	                 STSci (DSS)
	                 ESO (DSS)
	            - Added TRDB acces delay (for SMOKA from inside of NAOJ).
	            - Added FC instruments of Seimei Telescope.
                    - AllSkyCamera access via http proxy (udeer testing).
                    - Status monitoring for the 3.8-m Seimei telescope.
	                 (under testing).

2021.06.14 ver4.3.4 - Changed the limit size of all sky images to 1200 pix.
	            - Automatic rescale of X, Y, and d of all sky iamges over
	              the limit size.
                    - Adjusted all sky camera parameters.

2021.06.11 ver4.3.3 - A bug fix on memory release for broken all sky images.

2021.06.11 ver4.3.2 - Added Subaru's All Sky Image
	              (just for testing / Subaru inside only).
                     - A bug fix for all sky image flipping.
	             - Automatic broken image check for all sky images.

2021.04.27 ver4.3.1 - Fixed a bug in XMLRPC mode.

2021.04.06 ver4.3.0 - Added IRD and Seimei TriCCS for FC.
                    - Move AD, Tel Speed parameters to Obs position.

2021.03.23 ver4.2.0 - Revised Atmospherc Dispersion mappng, added KOOLS.
                    - Revised FC for KOOLS IFU.
	            - Read/Write for Seimei target list (.dat) file.

2021.01.25 ver4.1.0 - Added telescope azimuth mode (N=0 or S=0deg).
                    - Adjusted OAO sky camera parameters.
	            - Revised LAMOST FCDB search. Now supported DR5, DR6(low)
                      and DR6(Med).
                    - Added Subaru CHARIS, IRD, SWIMS, MIMIZUKU, OAO MuSCAT,
	              KOOLS-IFU, and Nishi-Harima NIC for SMOKA archive search.
                    - Updated SDSS data server to DR16.

2020.09.24 ver4.0.6 - Bug fix for TSC files auto loading.
                    - PMRA & PMDEC read for IRD OPE files.

2020.09.07 ver4.0.5 - Changed ASIVA's address to https.
                    - Revised https access for allsky cam images.

2020.08.18 ver4.0.4 - Added SWIMS into the FC instruments
	              PA is temporary (N-up = 0 and should be fixed
	              in observations).

2020.01.29 ver4.0.3 - Bug fix for multi-thread operation.
	              Added "return(NULL)" at the end of thread funcs.

2020.01.03 ver4.0.2 - Added Transient Name Server in Addobj search & WWWDB.

2019.12.26 ver4.0.1 - Revised downloading popup messages in FCDB search.
                    - Added g_thread_join() after threads finished.

2019.12.24 ver4.0.0 - Bug fix on target addtion from FCDB to Main Target list.
                    - Added "PM" button in Object LIst to search proper
	              motions of targets.
	            - Added SSL access mode to get All Sky Camera image.
                    - Revised code to set Obs/AllSky Cam parameters.
                    - Revised all multi-thread process into glib based ones.

2019.11.27 ver3.8.6 - Re-Added old PanSTARRS server in StSci (as default).
                      (MAST server is slow and unstable with large FOVs).
                    - Updated GSC catalog v2.3 --> v2.4.1 (for JWST)

2019.11.26 ver3.8.5 - Added UCAC4 for FCDB and MagDB.
                    - Updated PanSTARRS-1 server to the latest MAST server.

2019.11.06 ver3.8.4 - Added Non-Sidereal Check in HSC style OPE files.
                    - Revised to get Gen2 status server address from
                        [1] command line option "-s"
	                [2] environment variable "GEN2_RO_SERVER"
                       (priority [1] > [2])
                    - Reomoved the status server address setup in "~/.hskymon".

2019.11.04 ver3.8.3 - Revised Non-sidereal orbit plot in FC.
 	            - Automatic loading of Non-Sidereal tracking files
	              included in OPE files. NST files must be presented in
	              the same directory where OPE files are stored.
	            - Added text colors for RA & Dec of Non-sidereal targets
	              in Object List.

2019.04.09 ver3.8.2 - Bug fix for Gtk+3 ver.
                      removed editable=FALSE for spin_button.
                      Fixed inappropriate default font name.
                    - Changed xml tags for NED following server's side changes.
                    - Changed USNO-B1.0 catalog server to VizieR.
	            - Added proper motions in addobj_dialog.

2019.03.31 ver3.8.1 - Database updated SDSS DR14 --> DR15 .
	            - Markup text style is applied to label and
	              tree_view_column.
                    - Added
	                gtkut.c   gtkut.h
                    - Revised SSL_connect() in http-client.c

2019.03.11 ver3.8.0 - LGS PAM (satellite collision data) plot functions for
	              IRCS obs (imported fromhoe).
                      In hskymon PAM data can be imported before/after object
                      list loading. In hoe PAM must be imported only after
	              otherwise.

hsc.h
//    hskymon  from HDS OPE file Editor
//          New SkyMonitor for Subaru Gen2
//      hsc.h  --- Hyper Suprime Cam CCD chips information
//   
//                                           2012.10.22  A.Tajitsu

/*
[VERSION]
version: 116 CCDs for Summit4
comment: 2014/03/06 Updated by Utsumi, Y.
original: 20140206.cfg	# which file is the original one of this cfg file
[HEADER]
# row number
# key
# comment
# format f,	s, i, b: float, string, integer, bool
# separator ",\t"
*0:	T_GAIN1,	"Gain for channel 1",	f
*1:	T_GAIN2,	"Gain for channel 2",	f
*2:	T_GAIN3,	"Gain for channel 3",	f
*3:	T_GAIN4,	"Gain for channel 4",	f
(4:	T_CCDID,	"Name of CCD",	s 
*5:	T_XFLIP,	"CCD readout is x-flipped when create image",	b 
*6:	T_YFLIP,	"CCD readout is y-flipped when create image",	b 
7:	T_CCDSN,	"CCD Serial Number",   s
*8:	CRVAL1,		"Physical value of the reference pixel X",	f  =0.000
*9:	CRVAL2,		"Physical value of the reference pixel Y",	f  =0.000
10:	CRPIX1,		"Reference pixel in X (pixel)",	f
11:	CRPIX2,		"Reference pixel in Y (pixel)",	f
*12:	CDELT1,		"X Scale projected on detector (#/pix)",	f  =0.015
*13:	CDELT2,		"Y Scale projected on detector (#/pix)",	f  =0.015
*14:	CTYPE1,		"Pixel coordinate system",	s
*15:	CTYPE2,		"Pixel coordinate system",	s
*16:	CUNIT1,		"Units used in both CRVAL1 and CDELT1",	s
*17:	CUNIT2,		"Units used in both CRVAL2 and CDELT2",	s
18:	CD1_1,		"Pixel Coordinate translation matrix", f
19:	CD1_2,		"Pixel Coordinate translation matrix", f
20:	CD2_1,		"Pixel Coordinate translation matrix", f
21:	CD2_2,		"Pixel Coordinate translation matrix", f
22:	DET-ID,		"Sequential number assgined for the entire CCDs", i
*23:	T_NFRAME,	"Number of total frames",	i

typedef struct _HSCParam HSCParam;
struct _HSCParam{
  gint hsca;
  gint bees;
  
  gint ccdsn;
  
  gfloat crpix1;
  gfloat crpix2;

  gfloat cd11;
  gfloat cd12;
  gfloat cd21;
  gfloat cd22;

  gint det_id;
};

*/

typedef struct _HSCParam HSCParam;
struct _HSCParam{
  gint hsca;
  gint bees;
  
  gint ccdsn;
  
  gfloat crpix1;
  gfloat crpix2;
  
  gfloat cd1_1;
  gfloat cd1_2;
  gfloat cd2_1;
  gfloat cd2_2;
  
  gint det_id;
};

#define HSC_CHIP_ALL 116

typedef struct _HSCdead HSCdead;
struct _HSCdead{
  gint hsca;
  gint bees;
  
  gint ch;

  gfloat crpix1;
  gfloat crpix2;
  
  gfloat cd1_1;
  gfloat cd1_2;
  gfloat cd2_1;
  gfloat cd2_2;
};

const HSCdead hsc_dead[] = { 
  {20, 0, 3, 5269.833,	  4128.500,	0.000,	-0.015,	-0.015,	0.000},
  {20, 0, 4, 5269.833,	  4128.500,	0.000,	-0.015,	-0.015,	0.000},
  {51, 0, 1, -15956.833,  8597.167,	0.000,	-0.015,	-0.015,	0.000},
  {9,  1, 4, -1098.167,	  8863.833,	0.000,	0.015,	0.015,	0.000},
  {47, 1, 1, -11711.500, 13332.500,	0.000,	0.015,	0.015,	0.000},
  {47, 1, 2, -11711.500, 13332.500,	0.000,	0.015,	0.015,	0.000},
  {47, 1, 3, -11711.500, 13332.500,	0.000,	0.015,	0.015,	0.000},
  {47, 1, 4, -11711.500, 13332.500,	0.000,	0.015,	0.015,	0.000},
  {51, 1, 1, -15956.833,  8863.833,	0.000,	0.015,	0.015,	0.000},
  {51, 1, 2, -15956.833,  8863.833,	0.000,	0.015,	0.015,	0.000},
  {53, 1, 3, -13834.167, -4542.167,	0.000,	0.015,	0.015,	0.000}
};

#define HSC_DEAD_ALL 11

const HSCParam hsc_chips[] = {
   {0,	0,	99,		 -5343.500,	  4128.500,		0.000,	-0.015,	-0.015,	0.000,	73},
   {1,	0,	56,		 -5343.500,	  8597.167,		0.000,	-0.015,	-0.015,	0.000,	72},
   {2,	0,	118,		 -5343.500,	 13065.833,		0.000,	-0.015,	-0.015,	0.000,	71},
   {3,	0,	133,		 -5343.500,	 17534.500,		0.000,	-0.015,	-0.015,	0.000,	70},
   {4,	0,	73,		 -3220.833,	  4128.500,		0.000,	-0.015,	-0.015,	0.000,	65},
   {5,	0,	174,		 -3220.833,	  8597.167,		0.000,	-0.015,	-0.015,	0.000,	64},
   {6,	0,	59,		 -3220.833,	 13065.833,		0.000,	-0.015,	-0.015,	0.000,	63},
   {7,	0,	125,		 -3220.833,	 17534.500,		0.000,	-0.015,	-0.015,	0.000,	62},
   {8,	0,	160,		 -1098.167,	  4128.500,		0.000,	-0.015,	-0.015,	0.000,	57},
   {9,	0,	32,		 -1098.167,	  8597.167,		0.000,	-0.015,	-0.015,	0.000,	56},
   {10,	0,	52,		 -1098.167,	 13065.833,		0.000,	-0.015,	-0.015,	0.000,	55},
   {11,	0,	69,		 -1098.167,	 17534.500,		0.000,	-0.015,	-0.015,	0.000,	54},
   {12,	0,	127,		  1024.500,	  4128.500,		0.000,	-0.015,	-0.015,	0.000,	49},
   {13,	0,	21,		  1024.500,	  8597.167,		0.000,	-0.015,	-0.015,	0.000,	48},
   {14,	0,	148,		  1024.500,	 13065.833,		0.000,	-0.015,	-0.015,	0.000,	47},
   {15,	0,	103,		  1024.500,	 17534.500,		0.000,	-0.015,	-0.015,	0.000,	46},
   {16,	0,	168,		  3147.167,	  4128.500,		0.000,	-0.015,	-0.015,	0.000,	41},
   {17,	0,	173,		  3147.167,	  8597.167,		0.000,	-0.015,	-0.015,	0.000,	40},
   {18,	0,	178,		  3147.167,	 13065.833,		0.000,	-0.015,	-0.015,	0.000,	39},
   {19,	0,	43,		  3147.167,	 17534.500,		0.000,	-0.015,	-0.015,	0.000,	38},
   {20,	0,	48,		  5269.833,	  4128.500,		0.000,	-0.015,	-0.015,	0.000,	33},
   {21,	0,	58,		  5269.833,	  8597.167,		0.000,	-0.015,	-0.015,	0.000,	32},
   {22,	0,	136,		  5269.833,	 13065.833,		0.000,	-0.015,	-0.015,	0.000,	31},
   {23,	0,	38,		  5269.833,	 17534.500,		0.000,	-0.015,	-0.015,	0.000,	30},
   {24,	0,	162,		  7392.500,	  4128.500,		0.000,	-0.015,	-0.015,	0.000,	25},
   {25,	0,	34,		  7392.500,	  8597.167,		0.000,	-0.015,	-0.015,	0.000,	24},
   {26,	0,	76,		  7392.500,	 13065.833,		0.000,	-0.015,	-0.015,	0.000,	23},
   {27,	0,	102,		  7392.500,	 17534.500,		0.000,	-0.015,	-0.015,	0.000,	22},
   {28,	0,	35,		  9515.167,	  4128.500,		0.000,	-0.015,	-0.015,	0.000,	18},
   {29,	0,	128,		  9515.167,	  8597.167,		0.000,	-0.015,	-0.015,	0.000,	17},
   {30,	0,	149,		  9515.167,	 13065.833,		0.000,	-0.015,	-0.015,	0.000,	16},
   {31,	0,	143,		-13308.500,	 11691.167,		0.015,	0.000,	0.000,	-0.015,	100},
   {32,	0,	134,		 -7466.167,	  4128.500,		0.000,	-0.015,	-0.015,	0.000,	80},
   {33,	0,	65,		 -7466.167,	  8597.167,		0.000,	-0.015,	-0.015,	0.000,	79},
   {34,	0,	158,		 -7466.167,	 13065.833,		0.000,	-0.015,	-0.015,	0.000,	78},
   {35,	0,	74,		 15357.500,	 11691.167,		-0.015,	0.000,	0.000,	0.015,	102},
   {36,	0,	122,		 -9588.833,	 -9277.500,		0.000,	-0.015,	-0.015,	0.000,	89},
   {37,	0,	180,		 -9588.833,	 -4808.833,		0.000,	-0.015,	-0.015,	0.000,	88},
   {38,	0,	36,		 -9588.833,	  -340.167,		0.000,	-0.015,	-0.015,	0.000,	87},
   {39,	0,	176,		 -9588.833,	  4128.500,		0.000,	-0.015,	-0.015,	0.000,	86},
   {40,	0,	101,		 -9588.833,	  8597.167,		0.000,	-0.015,	-0.015,	0.000,	85},
   {41,	0,	30,		 -9588.833,	 13065.833,		0.000,	-0.015,	-0.015,	0.000,	84},
   {42,	0,	85,		-11711.500,	 -9277.500,		0.000,	-0.015,	-0.015,	0.000,	95},
   {43,	0,	172,		-11711.500,	 -4808.833,		0.000,	-0.015,	-0.015,	0.000,	94},
   {44,	0,	41,		-11711.500,	  -340.167,		0.000,	-0.015,	-0.015,	0.000,	93},
   {45,	0,	140,		-11711.500,	  4128.500,		0.000,	-0.015,	-0.015,	0.000,	92},
   {46,	0,	33,		-11711.500,	  8597.167,		0.000,	-0.015,	-0.015,	0.000,	91},
   {47,	0,	155,		-11711.500,	 13065.833,		0.000,	-0.015,	-0.015,	0.000,	90},
   {48,	0,	109,		-15956.833,	 -4808.833,		0.000,	-0.015,	-0.015,	0.000,	111},
   {51,	0,	113,		-15956.833,	  8597.167,		0.000,	-0.015,	-0.015,	0.000,	110},
   {52,	0,	114,		-13834.167,	 -9277.500,		0.000,	-0.015,	-0.015,	0.000,	109},
   {53,	0,	141,		-13834.167,	 -4808.833,		0.000,	-0.015,	-0.015,	0.000,	99},
   {54,	0,	146,		-13834.167,	  -340.167,		0.000,	-0.015,	-0.015,	0.000,	98},
   {55,	0,	95,		-13834.167,	  4128.500,		0.000,	-0.015,	-0.015,	0.000,	97},
   {56,	0,	96,		-13834.167,	  8597.167,		0.000,	-0.015,	-0.015,	0.000,	96},
   {57,	0,	108,		-13834.167,	 13065.833,		0.000,	-0.015,	-0.015,	0.000,	108},
   //[BEES1]
   {0,	1,	88,		 -5343.500,	  4395.167,		0.000,	0.015,	0.015,	0.000,	26},
   {1,	1,	94,		 -5343.500,	  8863.833,		0.000,	0.015,	0.015,	0.000,	27},
   {2,	1,	70,		 -5343.500,	 13332.500,		0.000,	0.015,	0.015,	0.000,	28},
   {3,	1,	157,		 -5343.500,	 17801.167,		0.000,	0.015,	0.015,	0.000,	29},
   {4,	1,	169,		 -3220.833,	  4395.167,		0.000,	0.015,	0.015,	0.000,	34},
   {5,	1,	171,		 -3220.833,	  8863.833,		0.000,	0.015,	0.015,	0.000,	35},
   {6,	1,	170,		 -3220.833,	 13332.500,		0.000,	0.015,	0.015,	0.000,	36},
   {7,	1,	61,		 -3220.833,	 17801.167,		0.000,	0.015,	0.015,	0.000,	37},
   {8,	1,	165,		 -1098.167,	  4395.167,		0.000,	0.015,	0.015,	0.000,	42},
   {9,	1,	49,		 -1098.167,	  8863.833,		0.000,	0.015,	0.015,	0.000,	43},
   {10,	1,	91,		 -1098.167,	 13332.500,		0.000,	0.015,	0.015,	0.000,	44},
   {11,	1,	119,		 -1098.167,	 17801.167,		0.000,	0.015,	0.015,	0.000,	45},
   {12,	1,	120,		  1024.500,	  4395.167,		0.000,	0.015,	0.015,	0.000,	50},
   {13,	1,	45,		  1024.500,	  8863.833,		0.000,	0.015,	0.015,	0.000,	51},
   {14,	1,	86,		  1024.500,	 13332.500,		0.000,	0.015,	0.015,	0.000,	52},
   {15,	1,	121,		  1024.500,	 17801.167,		0.000,	0.015,	0.015,	0.000,	53},
   {16,	1,	175,		  3147.167,	  4395.167,		0.000,	0.015,	0.015,	0.000,	58},
   {17,	1,	54,		  3147.167,	  8863.833,		0.000,	0.015,	0.015,	0.000,	59},
   {18,	1,	98,		  3147.167,	 13332.500,		0.000,	0.015,	0.015,	0.000,	60},
   {19,	1,	71,		  3147.167,	 17801.167,		0.000,	0.015,	0.015,	0.000,	61},
   {20,	1,	138,		  5269.833,	  4395.167,		0.000,	0.015,	0.015,	0.000,	66},
   {21,	1,	44,		  5269.833,	  8863.833,		0.000,	0.015,	0.015,	0.000,	67},
   {22,	1,	161,		  5269.833,	 13332.500,		0.000,	0.015,	0.015,	0.000,	68},
   {23,	1,	66,		  5269.833,	 17801.167,		0.000,	0.015,	0.015,	0.000,	69},
   {24,	1,	132,		  7392.500,	  4395.167,		0.000,	0.015,	0.015,	0.000,	74},
   {25,	1,	79,		  7392.500,	  8863.833,		0.000,	0.015,	0.015,	0.000,	75},
   {26,	1,	47,		  7392.500,	 13332.500,		0.000,	0.015,	0.015,	0.000,	76},
   {27,	1,	92,		  7392.500,	 17801.167,		0.000,	0.015,	0.015,	0.000,	77},
   {28,	1,	53,		  9515.167,	  4395.167,		0.000,	0.015,	0.015,	0.000,	81},
   {29,	1,	151,		  9515.167,	  8863.833,		0.000,	0.015,	0.015,	0.000,	82},
   {30,	1,	64,		  9515.167,	 13332.500,		0.000,	0.015,	0.015,	0.000,	83},
   {31,	1,	130,		-13575.167,	 11691.167,		-0.015,	0.000,	0.000,	0.015,	103},
   {32,	1,	57,		 -7466.167,	  4395.167,		0.000,	0.015,	0.015,	0.000,	19},
   {33,	1,	67,		 -7466.167,	  8863.833,		0.000,	0.015,	0.015,	0.000,	20},
   {34,	1,	117,		 -7466.167,	 13332.500,		0.000,	0.015,	0.015,	0.000,	21},
   {35,	1,	131,		 15624.167,	 11691.167,		0.015,	0.000,	0.000,	-0.015,	101},
   {36,	1,	90,		 -9588.833,	 -9010.833,		0.000,	0.015,	0.015,	0.000,	10},
   {37,	1,	31,		 -9588.833,	 -4542.167,		0.000,	0.015,	0.015,	0.000,	11},
   {38,	1,	179,		 -9588.833,	   -73.500,		0.000,	0.015,	0.015,	0.000,	12},
   {39,	1,	166,		 -9588.833,	  4395.167,		0.000,	0.015,	0.015,	0.000,	13},
   {40,	1,	46,		 -9588.833,	  8863.833,		0.000,	0.015,	0.015,	0.000,	14},
   {41,	1,	72,		 -9588.833,	 13332.500,		0.000,	0.015,	0.015,	0.000,	15},
   {42,	1,	89,		-11711.500,	 -9010.833,		0.000,	0.015,	0.015,	0.000,	4},
   {43,	1,	163,		-11711.500,	 -4542.167,		0.000,	0.015,	0.015,	0.000,	5},
   {44,	1,	159,		-11711.500,	   -73.500,		0.000,	0.015,	0.015,	0.000,	6},
   {45,	1,	150,		-11711.500,	  4395.167,		0.000,	0.015,	0.015,	0.000,	7},
   {46,	1,	37,		-11711.500,	  8863.833,		0.000,	0.015,	0.015,	0.000,	8},
   {47,	1,	156,		-11711.500,	 13332.500,		0.000,	0.015,	0.015,	0.000,	9},
   {48,	1,	107,		-15956.833,	 -4542.167,		0.000,	0.015,	0.015,	0.000,	104},
   {51,	1,	110,		-15956.833,	  8863.833,		0.000,	0.015,	0.015,	0.000,	105},
   {52,	1,	111,		-13834.167,	 -9010.833,		0.000,	0.015,	0.015,	0.000,	106},
   {53,	1,	126,		-13834.167,	 -4542.167,		0.000,	0.015,	0.015,	0.000,	0},  
   {54,	1,	39,		-13834.167,	   -73.500,		0.000,	0.015,	0.015,	0.000,	1},  
   {55,	1,	100,		-13834.167,	  4395.167,		0.000,	0.015,	0.015,	0.000,	2},
   {56,	1,	123,		-13834.167,	  8863.833,		0.000,	0.015,	0.015,	0.000,	3},
   {57,	1,	106,		-13834.167,	 13332.500,		0.000,	0.015,	0.015,	0.000,	107},
//[BEES2]	# for AG chip
   {1,	2,	116,		-15956.833,	   -73.500,		0.000,	0.015,	0.015,	0.000,  112},
   {2,	2,	105,		-15956.833,	  4395.167,		0.000,	0.015,	0.015,	0.000,	113},
   {17,	2,	80,		-15956.833,	  -340.167,		0.000,	-0.015,	-0.015,	0.000,	115},
   {18,	2,	135,		-15956.833,	  4128.500,		0.000,	-0.015,	-0.015,	0.000,	114}
};
